File: Python/Easy/2239_find_closest_number_to_zero.py
from typing import List

class Solution:
    """
    Given an integer array nums of size n, return the number with the value 
    closest to 0 in nums. If there are multiple answers, return the number 
    with the largest value.

    

    Example 1:

    Input: nums = [-4,-2,1,4,8]
    Output: 1
    Explanation:
    The distance from -4 to 0 is |-4| = 4.
    The distance from -2 to 0 is |-2| = 2.
    The distance from 1 to 0 is |1| = 1.
    The distance from 4 to 0 is |4| = 4.
    The distance from 8 to 0 is |8| = 8.
    Thus, the closest number to 0 in the array is 1.
    Example 2:

    Input: nums = [2,-1,1]
    Output: 1
    Explanation: 1 and -1 are both the closest numbers to 0, so 1 being larger is returned.
    

    Constraints:

    1 <= n <= 1000
    -105 <= nums[i] <= 105
    """
    def findClosestNumber_v1(self, nums: List[int]) -> int:
        """
        Original solution using a loop with conditionals.
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        closest: int = nums[0]
        for num in nums:
            val: int = abs(num)
            if val < abs(closest):
                closest = num
            elif val == abs(closest):
                closest = val if num > closest else closest
        return closest

    def findClosestNumber_map(self, nums: List[int]) -> int:
        """
        Optimized solution using min with a custom key.
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        return min(nums, key=lambda x: (abs(x), -x))

    def findClosestNumber_v2(self, nums: List[int]) -> int:
        """
        Solution with a simplified conditional approach.
        Time Complexity: O(n) for the loop + O(n) for the search = O(n + n) = O(2n) = O(n)
        Space Complexity: O(1)
        """
        closest = nums[0]
        for x in nums:
            if abs(x) < abs(closest):
                closest = x
        
        if closest < 0 and abs(closest) in nums:
            return abs(closest)
        else:
            return closest

if __name__ == "__main__":
    # Test cases
    test_cases = [
        ([-4, -2, 1, 4, 8], 1),
        ([2, -1, 1], 1),
        ([-10, -5, 5, 10], 5),
        ([0, -1, 1], 0),
        ([-2, -3, 3, 2], 2)
    ]
    
    solution = Solution()

    for nums, expected in test_cases:
        assert solution.findClosestNumber_v1(nums) == expected, f"v1 failed on {nums}"
        assert solution.findClosestNumber_map(nums) == expected, f"v2 failed on {nums}"
        assert solution.findClosestNumber_v2(nums) == expected, f"v3 failed on {nums}"
    
    print("All test cases passed!")

File: Golang/Easy/2239_find_closest_number_to_zero.go
package main

import (
	"fmt";
	"cmp";
	"slices";
)

// Solution groups all approaches for 2239. Find Closest Number to Zero.
//
// Given an integer array nums of size n, return the number closest to 0 in
// nums. If there are multiple answers, return the number with the largest value.
type Solution struct{}

// findClosestNumber returns the element in nums with the smallest absolute
// value, breaking ties in favour of the positive value.
//
// Single pass: tracks the best candidate seen so far. The abs helper is
// defined as a closure to avoid shadowing the built-in. Updates closest when
// a strictly smaller absolute value is found, or when the absolute values are
// equal but the current element is larger (positive beats negative on a tie).
//
// Time complexity:  O(n)
// Space complexity: O(1)
func (s Solution) findClosestNumber(nums []int) int {
	closest := nums[0]
	abs := func(n int) int {
		if n < 0 {
			return -n
		}
		return n
	}
	for _, num := range nums {
		val := abs(num)
		if val < abs(closest) {
			closest = num
		} else if val == abs(closest) && num > closest {
			closest = num
		}
	}
	return closest
}

// findClosestNumberMap returns the element in nums closest to zero using
// slices.MinFunc — the idiomatic Go equivalent of Python's min(key=...).
//
// slices.MinFunc accepts a comparator func(a, b T) int that must return:
//   - negative if a should be considered smaller (wins)
//   - zero     if a and b are equal
//   - positive if b should be considered smaller (wins)
//
// Use cmp.Compare(x, y) to produce a negative/zero/positive integer for
// ordered types instead of writing the comparison manually. The exact value
// is unspecified — only the sign matters to the comparator contract.
//
// When to use slices.MinFunc:
//   - Production code or take-home assignments — it is expressive and safe.
//   - Interviews where you can use the stdlib freely (some allow it).
//
// When NOT to use slices.MinFunc:
//   - LeetCode submissions — slices/cmp packages require Go 1.21+; older
//     judge versions will reject the import.
//   - Interviews where you are expected to demonstrate the algorithm itself
//     rather than delegate to a helper.
//
// Requires: "cmp" and "slices" imports (Go 1.21+).
//
// Time complexity:  O(n)
// Space complexity: O(1)
func (s Solution) findClosestNumberMap(nums []int) int {
	abs := func(n int) int {
		if n < 0 {
			return -n
		}
		return n
	}
	return slices.MinFunc(nums, func(a, b int) int {
		absA, absB := abs(a), abs(b)
		if absA != absB {
			return cmp.Compare(absA, absB)
		}
		return cmp.Compare(b, a) // larger value wins on tie
	})
}

func main() {
	testCases := []struct {
		nums     []int
		expected int
	}{
		{nums: []int{-4, -2, 1, 4, 8}, expected: 1},
		{nums: []int{2, -1, 1}, expected: 1},
		{nums: []int{-10, -5, 5, 10}, expected: 5},
		{nums: []int{0, -1, 1}, expected: 0},
		{nums: []int{-2, -3, 3, 2}, expected: 2},
	}

	sol := Solution{}

	for _, approach := range []struct {
		label string
		fn    func([]int) int
	}{
		{"findClosestNumber   ", sol.findClosestNumber},
		{"findClosestNumberMap", sol.findClosestNumberMap},
	} {
		fmt.Printf("--- %s ---\n", approach.label)
		for i, tc := range testCases {
			result := approach.fn(tc.nums)
			status := "PASS"
			if result != tc.expected {
				status = "FAIL"
			}
			fmt.Printf("  Example %d | Input: nums=%v | Expected: %d | Result: %d | [%s]\n",
				i+1, tc.nums, tc.expected, result, status)
		}
	}
}
